morpheus/_lib/src/messages/memory/tensor_memory.cpp

```diff
@@ -17,30 +17,79 @@
 
 #include "morpheus/messages/memory/tensor_memory.hpp"
 
+#include "morpheus/utilities/cupy_util.hpp"
+
+#include <pybind11/pybind11.h>  // for key_error & object
+
 #include <string>
 #include <vector>
 
 namespace morpheus {
 /****** Component public implementations *******************/
 /****** TensorMemory****************************************/
 TensorMemory::TensorMemory(size_t count) : count(count) {}
-TensorMemory::TensorMemory(size_t count, tensor_map_t &&tensors) : count(count), tensors(std::move(tensors)) {}
+TensorMemory::TensorMemory(size_t count, CupyUtil::tensor_map_t&& tensors) : count(count), tensors(std::move(tensors))
+{}
 
-bool TensorMemory::has_tensor(const std::string &name) const
+bool TensorMemory::has_tensor(const std::string& name) const
 {
     return this->tensors.find(name) != this->tensors.end();
 }
 
-TensorMemory::tensor_map_t TensorMemory::copy_tensor_ranges(
-    const std::vector<std::pair<TensorIndex, TensorIndex>> &ranges, size_t num_selected_rows) const
+CupyUtil::tensor_map_t TensorMemory::copy_tensor_ranges(const std::vector<std::pair<TensorIndex, TensorIndex>>& ranges,
+                                                        size_t num_selected_rows) const
 {
-    tensor_map_t tensors;
-    for (const auto &p : this->tensors)
+    CupyUtil::tensor_map_t tensors;
+    for (const auto& p : this->tensors)
     {
         tensors.insert(std::pair{p.first, p.second.copy_rows(ranges, num_selected_rows)});
     }
 
     return tensors;
 }
 
+/****** TensorMemoryInterfaceProxy *************************/
+std::shared_ptr<TensorMemory> TensorMemoryInterfaceProxy::init(std::size_t count, CupyUtil::py_tensor_map_t tensors)
+{
+    return std::make_shared<TensorMemory>(count, std::move(CupyUtil::cupy_to_tensors(tensors)));
+}
+
+std::size_t TensorMemoryInterfaceProxy::get_count(TensorMemory& self)
+{
+    return self.count;
+}
+
+CupyUtil::py_tensor_map_t TensorMemoryInterfaceProxy::get_tensors(TensorMemory& self)
+{
+    return CupyUtil::tensors_to_cupy(self.tensors);
+}
+
+void TensorMemoryInterfaceProxy::set_tensors(TensorMemory& self, CupyUtil::py_tensor_map_t tensors)
+{
+    self.tensors = std::move(CupyUtil::cupy_to_tensors(tensors));
+}
+
+const TensorObject& TensorMemoryInterfaceProxy::get_tensor_object(TensorMemory& self, const std::string& name)
+{
+    const auto tensor_itr = self.tensors.find(name);
+    if (tensor_itr == self.tensors.end())
+    {
+        throw pybind11::key_error{};
+    }
+
+    return tensor_itr->second;
+}
```

Comment:
This function seems out of place. I get the desire to reuse the pointer from `find()` but that is such a small optimization. It's better to have cleaner code here and remove this function. Then you can have the following for `get_tensor`:

```cpp
pybind11::object TensorMemoryInterfaceProxy::get_tensor(TensorMemory& self, const std::string name)
{
    if (!self.has_tensor(name)){
        throw pybind11::key_error{};
    }
    return CupyUtil::tensor_to_cupy(self.tensors[name]);
}
```